File: frontend/src/api/stats.ts
import client from './client'

export interface SummaryStats {
  total_spend: string
  transaction_count: number
  avg_per_month: string
  avg_per_day: string
}

export interface CategoryStat { category: string; total: string; count: number }
export interface MonthStat { month: string; total: string; count: number }
export interface DayStat { day: string; total: string; count: number }
export interface MerchantStat { merchant: string; total: string; count: number }

export interface StatFilters {
  category?: string
  start?: string
  end?: string
  search?: string
}

export const fetchSummary = (f: StatFilters) =>
  client.get<SummaryStats>('/stats/summary', { params: f }).then(r => r.data)

export const fetchByCategory = (f: StatFilters) =>
  client.get<CategoryStat[]>('/stats/by-category', { params: f }).then(r => r.data)

export const fetchByMonth = (f: StatFilters) =>
  client.get<MonthStat[]>('/stats/by-month', { params: f }).then(r => r.data)

export const fetchByDay = (f: StatFilters) =>
  client.get<DayStat[]>('/stats/by-day', { params: f }).then(r => r.data)

export const fetchTopMerchants = (f: StatFilters, limit = 10) =>
  client.get<MerchantStat[]>('/stats/top-merchants', { params: { ...f, limit } }).then(r => r.data)

export interface MonthSnapshot {
  month_name: string
  days_elapsed: number
  days_left: number
  current_total: string
  last_month_total: string
  pct_change: string
  top_category: string
  top_category_amount: string
  top_category_pct: string
  biggest_jump_category: string | null
  biggest_jump_amount: string
}

export const fetchMonthSnapshot = () =>
  client.get<MonthSnapshot>('/stats/month-snapshot').then(r => r.data)

File: backend/app/schemas/stats.py
from decimal import Decimal

from pydantic import BaseModel


class SummaryStats(BaseModel):
    total_spend: Decimal
    transaction_count: int
    avg_per_month: Decimal
    avg_per_day: Decimal


class CategoryStat(BaseModel):
    category: str
    total: Decimal
    count: int


class MonthStat(BaseModel):
    month: str  # "2025-01"
    total: Decimal
    count: int


class DayStat(BaseModel):
    day: str  # "2025-01-15"
    total: Decimal
    count: int


class MerchantStat(BaseModel):
    merchant: str
    total: Decimal
    count: int


class MonthSnapshot(BaseModel):
    month_name: str
    days_elapsed: int
    days_left: int
    current_total: Decimal
    last_month_total: Decimal
    pct_change: Decimal
    top_category: str
    top_category_amount: Decimal
    top_category_pct: Decimal
    biggest_jump_category: str | None
    biggest_jump_amount: Decimal
